espresso/streamer.go

```diff
@@ -58,49 +53,73 @@ type EspressoStreamer struct {
 	confirmedBatchPos uint64
 	// Hotshot block corresponding to the last safe batch
 	confirmedHotShotPos uint64
+	finalizedL1         eth.L1BlockRef
 
 	// Maintained in sorted order, but may be missing batches if we receive
 	// any out of order.
-	BatchBuffer BatchBuffer
+	BatchBuffer BatchBuffer[B]
+
+	UnmarshalBatch func([]byte) (*B, error)
+}
+
+func NewEspressoStreamer[B Batch](
+	namespace uint64,
+	l1Client L1Client,
+	espressoClient *espressoClient.Client,
+	lightClient *espressoLightClient.LightClientReader,
+	log log.Logger,
+	unmarshalBatch func([]byte) (*B, error),
+) EspressoStreamer[B] {
+	return EspressoStreamer[B]{
+		L1Client:            l1Client,
+		EspressoClient:      espressoClient,
+		EspressoLightClient: lightClient,
+		Log:                 log,
+
+		Namespace:   namespace,
+		BatchPos:    1,
+		BatchBuffer: NewBatchBuffer[B](),
+
+		UnmarshalBatch: unmarshalBatch,
+	}
 }
 
 // Reset the state to the last safe batch
-func (s *EspressoStreamer) Reset() {
+func (s *EspressoStreamer[B]) Reset() {
 	s.BatchPos = s.confirmedBatchPos + 1
 	s.hotShotPos = s.confirmedHotShotPos
-	s.BatchBuffer.Empty()
+	s.BatchBuffer.Clear()
 }
 
 // Handle both L1 reorgs and batcher restarts by updating our state in case it is
 // not consistent with what's on the L1. Returns true if the state was updated.
-func (s *EspressoStreamer) Refresh(ctx context.Context, syncStatus *eth.SyncStatus) (bool, error) {
+func (s *EspressoStreamer[B]) Refresh(ctx context.Context, syncStatus *eth.SyncStatus) (bool, error) {
 	if s.confirmedBatchPos == syncStatus.SafeL2.Number {
 		return false, nil
 	}
 
-	hotshotState, err := s.EspressoLightClient.LightClient.
-		FinalizedState(&bind.CallOpts{BlockNumber: new(big.Int).SetUint64(syncStatus.SafeL2.L1Origin.Number)})
-	if err != nil {
-		return false, err
-	}
+	//	hotshotState, err := s.EspressoLightClient.LightClient.
```

Comment:
Light client still doesn't work in tests, so for now I just set confirmedHotShotPos to 0 here. Could be either replaced by fixed light client or by manually tracking Espresso blocks we've seen that have corresponding SafeL2 as @Ayiga proposed

Reply:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478924?focus=true